### ISSUE TITLE
RDCatch: Segfault on one-shot MACRO

### ISSUE THREAD

**Issue by alexolivan** [ElvishArtisan/rivendell#782]
Hi.

Something strange happens in rdcatch after the latest commit solving playback segfault.
First, I scheduled some 'one shot' playbacks that executed fine.

Then I sceduled the whole 24 Top of Hour GTS pips, Mon to Sun, along with some MACROS... Soon I'm sure one of the playbacks didn't work... so I tailed to console syslog.

As MACROS had to execute, I saw no activity at all on the syslog, and no action on LogMachines affected by those macros, like if they were totally ignored.
So, I created a 'one shot' macro (just a TA 0!) and  rdcatchd segfaulted when it executed it.
Here's syslog output where event creation 207 is created pointing to MACRO 50662, and how it crashed upon executing it:

```
Mar  3 17:05:04 debian ripcd: sent notification: "NOTIFY CATCH_EVENT ADD 207" to 239.192.255.72:20539
Mar  3 17:05:04 debian rdcatchd: event 207 not found, not removed
Mar  3 17:05:04 debian rdcatchd: loading event 207, Type: macro, Cart: 50662
Mar  3 17:05:20 debian rdcatchd: clearing event_id: 207
Mar  3 17:05:20 debian kernel: [109755.046069] rdcatchd[146280]: segfault at 14 ip 00007f82c4c417f0 sp 00007ffd99c0ffc8 error 6 in libQt5Core.so.5.15.2[7f82c49e8000+2fe000]
Mar  3 17:05:20 debian kernel: [109755.046089] Code: 48 89 df 83 e2 fd 09 c2 88 53 1c c0 ea 02 83 e2 03 e8 94 53 ff ff 89 43 10 5b c3 90 66 66 2e 0f 1f 84 00 00 00 00 00 0f 1f 00 <89> 77 14 e9 a8 ff ff ff 0f 1f 84 00 00 00 00 00 8b 46 14 39 47 10
Mar  3 17:05:20 debian rdcatchd: purged event 207, Type: macro, Cart: 50662
Mar  3 17:05:20 debian rdcatchd: executing macro cart: 50662
Mar  3 17:05:20 debian caed: removed connection 21 [::ffff:127.0.0.1:57966]

```

The fun thing is that rdcatchd has been skipping/ignoring macros for a while doing playouts (I'm almost sure not all of them), until I tripped something with the 'one shot' macro.
I'll try to run this with dbg to see if I can provide more details on what's going on.

Cheers.  
 
UPDATE1:

I'm trying to reproduce the error but I'm unable to.
I'm running rdcatchd with dbg in a terminal as root, and I have syslog tailed on another.
To replicate yesterday scenario, I have rdcatch module opened, with absolutely no changes... so, the same schedule than yesterday appears as loaded from DB.
But unlike yesterday, macros are running with rdcatchd and ripcd activity written to syslog, so the macro lines can be nicely seen as received RMLs.
I will test if adding 'one shot' macros is the cause of the weird behaviour

UPDATE2:

Ok, it segfaulted as soon as a one-shot MACRO scheduled Cart is run (unfortunately I think there's no dbgsym package for libqt5core in bullseye for some reason) so I feel backtrace may not be as useful as it should...here it is anyways:

```
Thread 1 "rdcatchd" received signal SIGSEGV, Segmentation fault.
0x00007ffff639b7f0 in QTimer::start(int) ()
   from /lib/x86_64-linux-gnu/libQt5Core.so.5
(gdb) backtrace
#0  0x00007ffff639b7f0 in QTimer::start(int) ()
   from /lib/x86_64-linux-gnu/libQt5Core.so.5
#1  0x00007ffff79d3bfe in RDTimeEngine::StartEvent (this=0x555555611b90, 
    id=208) at rdtimeengine.cpp:112
#2  0x00007ffff63975e0 in ?? () from /lib/x86_64-linux-gnu/libQt5Core.so.5
#3  0x00007ffff6398c8e in QSignalMapper::mapped(int) ()
   from /lib/x86_64-linux-gnu/libQt5Core.so.5
#4  0x00007ffff639a15c in QSignalMapper::map(QObject*) ()
   from /lib/x86_64-linux-gnu/libQt5Core.so.5
#5  0x00007ffff63975e0 in ?? () from /lib/x86_64-linux-gnu/libQt5Core.so.5
#6  0x00007ffff639b4ba in QTimer::timeout(QTimer::QPrivateSignal) ()
   from /lib/x86_64-linux-gnu/libQt5Core.so.5
#7  0x00007ffff638cf2f in QObject::event(QEvent*) ()
   from /lib/x86_64-linux-gnu/libQt5Core.so.5
#8  0x00007ffff6360f9f in QCoreApplication::notifyInternal2(QObject*, QEvent*)
    () from /lib/x86_64-linux-gnu/libQt5Core.so.5
#9  0x00007ffff63b78e3 in QTimerInfoList::activateTimers() ()
   from /lib/x86_64-linux-gnu/libQt5Core.so.5
#10 0x00007ffff63b819c in ?? () from /lib/x86_64-linux-gnu/libQt5Core.so.5
#11 0x00007ffff4598e6b in g_main_context_dispatch ()
   from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#12 0x00007ffff4599118 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#13 0x00007ffff45991cf in g_main_context_iteration ()
--Type <RET> for more, q to quit, c to continue without paging--
   from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#14 0x00007ffff63b851f in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /lib/x86_64-linux-gnu/libQt5Core.so.5
#15 0x00007ffff635f98b in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /lib/x86_64-linux-gnu/libQt5Core.so.5
#16 0x00007ffff6367c00 in QCoreApplication::exec() ()
   from /lib/x86_64-linux-gnu/libQt5Core.so.5
#17 0x000055555555f830 in main ()
``` 

One thing but that I'm noticing is that rdcatch module seems to act weirdly when managing events regarding their sorting on the timeline... new events do always land on the bottom of the list, so you need to close and reopen rdcatch to see events in the right chronological order... probably unrelated... but I'm starting to feel that, once again, pesky qt5 things are causing some headache here.

**Comment by fpeyraud** [comment]
There is definitely something happening with rdcatchd causing it to crash or partially crash is some circumstances and I don't succeed to spot it clearly. It's causing a lot of trouble to some customers who trigger log generation via rdcatch tasks. They often ask to switch to cron tasks, even if the maintainance is more complicated. The good thing is that everything seem to work correctly after rdcatchd is restarted. The bad thing is that there is no way to make it restart except by restarting the whole rivendell service.

**Comment by ElvishArtisan** [comment]
> It's causing a lot of trouble to some customers who trigger log generation via rdcatch tasks.

Customers? I would highly recommend __not__ using v4.x for production workloads just yet.

That said, there has recently (within the last ten days or so) been numerous commits address overall stability in the RDCatch subsystem. Please consider giving the current HEAD for v4.x (branch 'qt5') do verify that these have all been addressed.

**Comment by alexolivan** [comment]
Update on this.

Notice that macro execution call in RDCatch is working for me (I have some problems with MD RML command in my macros, for some reason sometimes the un duck operation fails, but I assume those to be unrelated to rdcatchd)
The persisting problem is whit One-Shot Macro events leading to crash rdcatchd.

So, after updating Rivendell to latest commit, I have scheduled a freshly created One-Shot Macro Cart execution:

```
May  2 11:35:16 debian ripcd: sent notification: "NOTIFY CATCH_EVENT ADD 217" to 239.192.255.72:20539
May  2 11:35:16 debian rdcatchd: event 217 not found, not removed
May  2 11:35:16 debian rdcatchd: loading event 217, Type: macro, Cart: 50662
May  2 11:35:16 debian rdcatchd: removed event 217, Type: macro, Cart: 50662
May  2 11:35:16 debian rdcatchd: loading event 217, Type: macro, Cart: 50662
```

First thing I noticed is that duplicated events appeared in GUI:

![Screenshot at 2022-05-02 11-35-43](https://user-images.githubusercontent.com/1419896/166215757-da0ed976-d563-43aa-8b6f-1bbf61ac50bb.png)

Restarting RDcatch, as I expected, fixed the annoyance (indicating it was just front-end issue... back-end was successfully updated apparently)

![Screenshot at 2022-05-02 11-36-13](https://user-images.githubusercontent.com/1419896/166215881-ca29b2d7-0a2a-46ba-ae42-22b158351347.png)

However, as the execution moment was reached, a segfault appeared in syslog... I have no backtrace at hand :-(

```
May  2 11:37:00 debian rdcatchd: clearing event_id: 217
May  2 11:37:00 debian rdcatchd: purged event 217, Type: macro, Cart: 50662
May  2 11:37:00 debian rdcatchd: executing macro cart: 50662
May  2 11:37:00 debian kernel: [127547.557227] rdcatchd[190684]: segfault at 14 ip 00007f089846a7f0 sp 00007ffebdb42f18 error 6 in libQt5Core.so.5.15.2[7f0898211000+2fe000]
May  2 11:37:00 debian kernel: [127547.557255] Code: 48 89 df 83 e2 fd 09 c2 88 53 1c c0 ea 02 83 e2 03 e8 94 53 ff ff 89 43 10 5b c3 90 66 66 2e 0f 1f 84 00 00 00 00 00 0f 1f 00 <89> 77 14 e9 a8 ff ff ff 0f 1f 84 00 00 00 00 00 8b 46 14 39 47 10
```

**Comment by ElvishArtisan** [comment]
Unable to reproduce here. What are the contents of cart 50662?

**Comment by alexolivan** [comment]
Sure! ... but nothing Fancy, just TA 0! :-)

![Screenshot at 2022-05-03 09-48-38](https://user-images.githubusercontent.com/1419896/166419981-cec597e9-3687-4b4a-9ff1-b8aa04685fa6.png)

**Comment by alexolivan** [comment]
Ok... today I repeated the test several times, consistenty crashing.
Cart to be used for the test is the same TA 0! cart (I have tested it indepndently from RDLibrary and it is harmless to run, it works)

Steps followed, Open rdcatch, create a new event, select Macro, mark the one-shot checkbox, mark today's dayweek to ensure execution, select Cart from dropdown, select a time in the future, save... and wait.

Tested once to have rdcatchd crashed, then ran it from GDB on console and repeat.
Here's resulting backtrace:

```
Thread 1 "rdcatchd" received signal SIGSEGV, Segmentation fault.
0x00007ffff63997f0 in QTimer::start(int) ()
   from /lib/x86_64-linux-gnu/libQt5Core.so.5
(gdb) backtrace
#0  0x00007ffff63997f0 in QTimer::start(int) ()
   from /lib/x86_64-linux-gnu/libQt5Core.so.5
#1  0x00007ffff79d3b0e in RDTimeEngine::StartEvent (this=0x555555612560, id=221)
    at rdtimeengine.cpp:112
#2  0x00007ffff63955e0 in ?? () from /lib/x86_64-linux-gnu/libQt5Core.so.5
#3  0x00007ffff6396c8e in QSignalMapper::mapped(int) ()
   from /lib/x86_64-linux-gnu/libQt5Core.so.5
#4  0x00007ffff639815c in QSignalMapper::map(QObject*) ()
   from /lib/x86_64-linux-gnu/libQt5Core.so.5
#5  0x00007ffff63955e0 in ?? () from /lib/x86_64-linux-gnu/libQt5Core.so.5
#6  0x00007ffff63994ba in QTimer::timeout(QTimer::QPrivateSignal) ()
   from /lib/x86_64-linux-gnu/libQt5Core.so.5
#7  0x00007ffff638af2f in QObject::event(QEvent*) ()
   from /lib/x86_64-linux-gnu/libQt5Core.so.5
#8  0x00007ffff635ef9f in QCoreApplication::notifyInternal2(QObject*, QEvent*) ()
   from /lib/x86_64-linux-gnu/libQt5Core.so.5
#9  0x00007ffff63b58e3 in QTimerInfoList::activateTimers() ()
   from /lib/x86_64-linux-gnu/libQt5Core.so.5
#10 0x00007ffff63b619c in ?? () from /lib/x86_64-linux-gnu/libQt5Core.so.5
#11 0x00007ffff4596e6b in g_main_context_dispatch ()
   from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#12 0x00007ffff4597118 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#13 0x00007ffff45971cf in g_main_context_iteration ()
   from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#14 0x00007ffff63b651f in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /lib/x86_64-linux-gnu/libQt5Core.so.5
#15 0x00007ffff635d98b in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) ()
   from /lib/x86_64-linux-gnu/libQt5Core.so.5
#16 0x00007ffff6365c00 in QCoreApplication::exec() ()
--Type <RET> for more, q to quit, c to continue without paging--c
   from /lib/x86_64-linux-gnu/libQt5Core.so.5
#17 0x000055555555f8d0 in main (argc=<optimized out>, argv=<optimized out>) at rdcatchd.cpp:2586
```


Related syslog output:

```
May  5 15:27:44 debian rdcatchd: created connection 0
May  5 15:28:41 debian ripcd: sent notification: "NOTIFY CATCH_EVENT ADD 221" to 239.192.255.72:20539
May  5 15:28:41 debian rdcatchd: event 221 not found, not removed
May  5 15:28:41 debian rdcatchd: loading event 221, Type: macro, Cart: 50662
May  5 15:30:00 debian rdcatchd: clearing event_id: 221
May  5 15:30:00 debian rdcatchd: purged event 221, Type: macro, Cart: 50662
May  5 15:30:00 debian rdcatchd: executing macro cart: 50662
```
 
Hope this helps.
Regards.

**Comment by ElvishArtisan** [comment]
Thank you. _Very_ helpful!

**Comment by ElvishArtisan** [comment]
Still unable to reproduce it here, but on the basis of your stack trace, is likely fixed in e53242c. Please test!
